### DISCUSSION TITLE
[Lua] osc_plus 增强版控制界面（原 osc_lazy）

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#18] [INFO 信息]
# 本文已过时

参见 https://github.com/hooke007/MPV_lazy/commit/20acfc0fd93202606916289fa422a7bc405ba0c4

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

## 预览

[**osc_plus.lua**](https://github.com/hooke007/MPV_lazy/blob/2027fb8b2ec766896773c6803c9b7a33a4fc6f12/portable_config/scripts/osc_plus.lua) 是基于mpv原版OSC进行功能强化与界面美化的外置脚本。
在继承上游的全部特性的同时，进行了较大的改动。

普通脚本，使用常规的安装方式即可。  [【通用引导参考】](https://hooke007.github.io/unofficial/mpv_start.html#id14)
默认不内建预绑定的快捷键方案，可用的快捷键绑定 input.conf 示例：
https://github.com/hooke007/MPV_lazy/blob/c1aa59a5da0b59dab98d73d1b4a343870e13e3a1/portable_config/scripts/osc_plus.lua#L10-L14

继承了上游 [osc.conf](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/script-opts/osc.conf) 的全部选项外，还有更多的其它选项可供自定义，具体查看示范 **[osc_plus.conf](https://github.com/hooke007/MPV_lazy/blob/2027fb8b2ec766896773c6803c9b7a33a4fc6f12/portable_config/script-opts/osc_plus.conf)**

粗略配置后的运行时预览图：（在此配置文件中使用 `layout=bottombox` 即可启用预览图所示的布局）

<img width=600 
src="https://user-images.githubusercontent.com/41094733/132130429-71fc5d90-0011-466d-9ccf-4afecd2937b6.jpg">

## 特性说明

1. 启动mpv前可以不在 **mpv.conf** 中设置 `osc=no` 关闭内置OSC，但会影响个别功能。

2. 兼容适配新的缩略图引擎（没有下列附属脚本就自然不显示时间轴缩略图预览）
 - [thumbfast改版](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/thumbfast.lua) & 预设设置 [thumbfast.conf](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/script-opts/thumbfast.conf)

3. 新增布局bottombox：基于原始box布局改造，支持右侧 **动态次标题**（可自定义）；光标悬停时间轴时，自动默认切换为均衡器信息（可自定义）； **右侧新按钮**（联动stats.lua）：右键 临时显示统计页面4，滚轮上 临时显示统计页面1，滚轮下 临时显示统计页面2，左键 常驻/关闭（常驻状态下切换页面也为常驻显示）
<img src="https://user-images.githubusercontent.com/41094733/132130563-53fe4dfb-f8e3-4010-a05f-fffced77216f.jpg" width=500>

4. **音轨** 和 **字幕** 按钮支持 鼠标悬停时滚轮快速切换不同音轨/字幕（<s>原本只有 **音量** 按钮支持鼠标悬停时滚轮快速调整音量</s> 与原版的切换方向相反）
 **进度条** 支持 关键帧/逐秒/逐帧 三种不同模式的滚轮跳转（原版仅一种）
<img src="https://user-images.githubusercontent.com/41094733/132130489-75fc7802-bca1-4bec-a3f5-6864363bc4f7.jpg" width=400>

5. 无边框模式下 **顶部标题** 支持自定义
<img src="https://user-images.githubusercontent.com/41094733/132130556-1a80683f-ff06-4dd4-b475-e242bccbe1b8.jpg" width=800>

6. osc弹出文本的本地化（用灰色圈出的 按钮 原本就支持 **鼠标中键** // **shift** + **鼠标左键** 弹出显示这些OSD信息）
<img src="https://user-images.githubusercontent.com/41094733/132130672-61d80983-cc7f-4fb7-afec-17e8419fc627.jpg" width=500>

_更多特性和功能参见上方 osc_plus.conf 列出的新增用户选项_

**Comment by hooke007** [comment]
_已移除旧缩略图引擎的支持_

**Reply by lbcmiao** [reply]
进度条在按钮的下方，感觉很怪异，大多数播放器都是进度条在按钮上方的。

**Reply by hooke007** [reply]
原始box就是这样的，没有更改的打算

**Comment by zhongfly** [comment]
~~[thumbfast](https://github.com/po5/thumbfast)粗略尝试了下，修改成mpvnet可用的，应该是另外打开了一个mpv窗口在上方显示缩略图？~~

~~修改259行如下，即可在mpvnet上使用，效果不佳~~
```
            "mpvnet", path, "--no-config", "--msg-level=all=no", "--idle", "--pause", "--no-border",
```

**Reply by hooke007** [reply]
你确定这个能用？我无法复现，而且mpvnet不支持--o的编码指令 https://github.com/stax76/mpv.net/issues/387

**Reply by zhongfly** [reply]
还真能用…配合你的osc_lazy，的确出现了一个mpvnet窗口显示缩略图

**Reply by hooke007** [reply]
> 我无法复现

**Reply by zhongfly** [reply]
![image](https://user-images.githubusercontent.com/11155705/191539663-1efe1b1e-993e-45c5-b00f-b35b0885d8d7.png)

**Reply by hooke007** [reply]
我只有报错，表示这是玄学事件
![Snipaste_2022-09-21_16-08-31](https://user-images.githubusercontent.com/41094733/191541487-5260703d-0346-431a-a79f-b1d5456a85cc.jpg)

**Reply by zhongfly** [reply]
你没有看到效果可能是因为缩略图的那个窗口没有被置顶，被挡住了
在mpv.net-6.0.3.1-beta上复现
[portable_config.zip](https://github.com/hooke007/MPV_lazy/files/9617949/portable_config.zip)

而且我的PATH里也没有设置mpv/mpv.net的路径（终端里输入mpv回车显示“不是内部或外部命令，也不是可运行的程序”），缩略图的那个窗口也在任务栏里也显示为mpv.net，不过不支持右键菜单

**Reply by hooke007** [reply]
不是置顶，是你允许了mpvnet多进程.（这是原因）
而且并非以脚本预期的方式在编码，只是在播放文件，正如我在那个issue里提到的错误行为
而且也未正确追踪时间轴，只是看起来像是运行起来了，实际完全不是（瞎猫撞到了死。。甚至不算耗子）

**Reply by hooke007** [reply]
你要正常跑也不用这么麻烦，直接目录里丢个mpv.exe进去就完事了

**Reply by zhongfly** [reply]
不过也可以通过在mpvnet目录里放一个mpv.exe和mpv.com来使用……就不用修改脚本了

**Reply by hooke007** [reply]
甚至不需要 mpv.com

**Comment by suixinex** [comment]
有个小问题，相比旧引擎（Thumbnailer）是否更推荐使用新引擎（thumbfast），或者说这两个有哪些优缺点呢，另外这个[thumbfast](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/thumbfast.lua)是个人修改的吗，原脚本没有precise和hwdec选项

**Reply by hooke007** [reply]
一个一次性生成全部，一个用时按需生成。新引擎还不够稳定，等上游bug修的上不多了会移除旧引擎。文中已写，另外此仓库的所有脚本都是修改版

**Reply by suixinex** [reply]
好的，了解了，谢谢解答
